diff --git a/include/boost/mpl/vector/vector100.hpp b/include/boost/mpl/vector/vector100.hpp
new file mode 100644
index 0000000..3343624
--- /dev/null
+++ b/include/boost/mpl/vector/vector100.hpp
@@ -0,0 +1,18 @@
+
+#ifndef BOOST_MPL_VECTOR_VECTOR100_HPP_INCLUDED
+#define BOOST_MPL_VECTOR_VECTOR100_HPP_INCLUDED
+
+#   include <boost/mpl/vector/vector50.hpp>
+#   include <boost/mpl/aux_/config/typeof.hpp>
+#   include <boost/mpl/aux_/config/ctps.hpp>
+#   include <boost/preprocessor/iterate.hpp>
+
+namespace boost { namespace mpl {
+
+#   define BOOST_PP_ITERATION_PARAMS_1 \
+    (3,(51, 100, <boost/mpl/vector/aux_/numbered.hpp>))
+#   include BOOST_PP_ITERATE()
+
+}}
+
+#endif // BOOST_MPL_VECTOR_VECTOR100_HPP_INCLUDED
diff --git a/include/boost/mpl/vector/vector100_c.hpp b/include/boost/mpl/vector/vector100_c.hpp
new file mode 100644
index 0000000..541c548
--- /dev/null
+++ b/include/boost/mpl/vector/vector100_c.hpp
@@ -0,0 +1,19 @@
+
+#ifndef BOOST_MPL_VECTOR_VECTOR100_C_HPP_INCLUDED
+#define BOOST_MPL_VECTOR_VECTOR100_C_HPP_INCLUDED
+
+#   include <boost/mpl/vector/vector50_c.hpp>
+#   include <boost/mpl/vector/vector100.hpp>
+#   include <boost/mpl/aux_/config/typeof.hpp>
+#   include <boost/mpl/aux_/config/ctps.hpp>
+#   include <boost/preprocessor/iterate.hpp>
+
+namespace boost { namespace mpl {
+
+#   define BOOST_PP_ITERATION_PARAMS_1 \
+    (3,(51, 100, <boost/mpl/vector/aux_/numbered_c.hpp>))
+#   include BOOST_PP_ITERATE()
+
+}}
+
+#endif // BOOST_MPL_VECTOR_VECTOR100_C_HPP_INCLUDED
