diff --git a/include/boost/gil/channel_algorithm.hpp b/include/boost/gil/channel_algorithm.hpp
index 12ecd01b0..d4f733e38 100644
--- a/include/boost/gil/channel_algorithm.hpp
+++ b/include/boost/gil/channel_algorithm.hpp
@@ -263,8 +263,8 @@ struct channel_converter_unsigned_integral_nondivisible<SrcChannelV, DstChannelV
     {
         using dest_t = typename base_channel_type<DstChannelV>::type;
         return DstChannelV(
-            static_cast<dest_t>(src * unsigned_integral_max_value<DstChannelV>::value)
-            / unsigned_integral_max_value<SrcChannelV>::value);
+            static_cast<dest_t>(src * unsigned_integral_max_value<DstChannelV>::value
+            / unsigned_integral_max_value<SrcChannelV>::value));
     }
 };
 
