diff --git a/include/boost/interprocess/detail/os_file_functions.hpp b/include/boost/interprocess/detail/os_file_functions.hpp
index 6ff4cb5..4e80ab6 100644
--- a/include/boost/interprocess/detail/os_file_functions.hpp
+++ b/include/boost/interprocess/detail/os_file_functions.hpp
@@ -684,53 +684,6 @@ inline bool release_file_lock_sharable(file_handle_t hnd)
 {  return 0 == ::flock(hnd, LOCK_UN); }
 #endif
 
-inline bool delete_subdirectories_recursive
-   (const std::string &refcstrRootDirectory, const char *dont_delete_this)
-{
-   DIR *d = opendir(refcstrRootDirectory.c_str());
-   if(!d) {
-      return false;
-   }
-
-   struct dir_close
-   {
-      DIR *d_;
-      dir_close(DIR *d) : d_(d) {}
-      ~dir_close() { ::closedir(d_); }
-   } dc(d); (void)dc;
-
-   struct ::dirent *de;
-   struct ::stat st;
-   std::string fn;
-
-   while((de=::readdir(d))) {
-      if( de->d_name[0] == '.' && ( de->d_name[1] == '\0'
-            || (de->d_name[1] == '.' && de->d_name[2] == '\0' )) ){
-         continue;
-      }
-      if(dont_delete_this && std::strcmp(dont_delete_this, de->d_name) == 0){
-         continue;
-      }
-      fn = refcstrRootDirectory;
-      fn += '/';
-      fn += de->d_name;
-
-      if(std::remove(fn.c_str())) {
-         if(::stat(fn.c_str(), & st)) {
-            return false;
-         }
-         if(S_ISDIR(st.st_mode)) {
-            if(!delete_subdirectories_recursive(fn, 0) ){
-               return false;
-            }
-         } else {
-            return false;
-         }
-      }
-   }
-   return std::remove(refcstrRootDirectory.c_str()) ? false : true;
-}
-
 template<class Function>
 inline bool for_each_file_in_dir(const char *dir, Function f)
 {
@@ -773,11 +726,6 @@ inline bool for_each_file_in_dir(const char *dir, Function f)
 }
 
 
-//This function erases all the subdirectories of a directory except the one pointed by "dont_delete_this"
-inline bool delete_subdirectories(const std::string &refcstrRootDirectory, const char *dont_delete_this)
-{
-   return delete_subdirectories_recursive(refcstrRootDirectory, dont_delete_this );
-}
 
 #endif   //#if defined (BOOST_INTERPROCESS_WINDOWS)
 
