diff --git a/include/boost/beast/websocket/detail/hybi13.ipp b/include/boost/beast/websocket/detail/hybi13.ipp
index 0cbab153..3360b398 100644
--- a/include/boost/beast/websocket/detail/hybi13.ipp
+++ b/include/boost/beast/websocket/detail/hybi13.ipp
@@ -47,11 +47,12 @@ make_sec_ws_accept(
     beast::detail::init(ctx);
     beast::detail::update(ctx, key.data(), key.size());
     beast::detail::update(ctx, guid.data(), guid.size());
-    char digest[beast::detail::sha1_context::digest_size];
+    std::string digest;
+    digest.resize(beast::detail::sha1_context::digest_size);
     beast::detail::finish(ctx, &digest[0]);
     accept.resize(accept.max_size());
     accept.resize(beast::detail::base64::encode(
-        accept.data(), &digest[0], sizeof(digest)));
+        accept.data(), &digest[0], digest.size()));
 }
 
 } // detail
