diff --git a/include/boost/dll/detail/elf_info.hpp b/include/boost/dll/detail/elf_info.hpp
index f65804b..6d336ed 100644
--- a/include/boost/dll/detail/elf_info.hpp
+++ b/include/boost/dll/detail/elf_info.hpp
@@ -228,7 +228,7 @@ public:
         ret.reserve(symbols.size());
         for (std::size_t i = 0; i < symbols.size(); ++i) {
             if (is_visible(symbols[i])) {
-                ret.push_back(&text[0] + symbols[i].st_name);
+                ret.emplace_back(text.begin() + symbols[i].st_name, text.end());
                 if (ret.back().empty()) {
                     ret.pop_back(); // Do not show empty names
                 }
@@ -276,7 +276,7 @@ public:
 
         for (std::size_t i = 0; i < symbols.size(); ++i) {
             if (symbols[i].st_shndx == index && is_visible(symbols[i])) {
-                ret.push_back(&text[0] + symbols[i].st_name);
+                ret.emplace_back(text.begin() + symbols[i].st_name, text.end());
                 if (ret.back().empty()) {
                     ret.pop_back(); // Do not show empty names
                 }
