diff --git a/include/boost/regex/v5/w32_regex_traits.hpp b/include/boost/regex/v5/w32_regex_traits.hpp
index 7e90e4ba..3dbd5c1f 100644
--- a/include/boost/regex/v5/w32_regex_traits.hpp
+++ b/include/boost/regex/v5/w32_regex_traits.hpp
@@ -29,8 +29,15 @@
 #include <boost/regex/v5/primary_transform.hpp>
 #include <boost/regex/v5/object_cache.hpp>
 
-#define VC_EXTRALEAN
-#define WIN32_LEAN_AND_MEAN
+#ifndef VC_EXTRALEAN
+#  define VC_EXTRALEAN
+#endif
+#ifndef WIN32_LEAN_AND_MEAN
+#  define WIN32_LEAN_AND_MEAN
+#endif
+#ifndef NOMINMAX
+#  define NOMINMAX
+#endif
 #include <windows.h>
 
 #if defined(_MSC_VER) && !defined(_WIN32_WCE) && !defined(UNDER_CE)
