diff --git a/tools/clang-format/CMakeLists.txt b/tools/clang-format/CMakeLists.txt
index 6a24e13..57ffd09 100644
--- a/tools/clang-format/CMakeLists.txt
+++ b/tools/clang-format/CMakeLists.txt
@@ -11,7 +11,8 @@ target_link_libraries(clang-format
   clangToolingCore
   )
 
-install(TARGETS clang-format RUNTIME DESTINATION bin)
+install(TARGETS clang-format EXPORT clang-format-targets RUNTIME DESTINATION bin)
+install(EXPORT clang-format-targets DESTINATION bin/cmake)
 install(PROGRAMS clang-format-bbedit.applescript DESTINATION share/clang)
 install(PROGRAMS clang-format-diff.py DESTINATION share/clang)
 install(PROGRAMS clang-format-sublime.py DESTINATION share/clang)
