diff --git a/CMakeLists.txt b/CMakeLists.txt
index c860dc4..ec2564e 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1,4 +1,4 @@
-cmake_minimum_required(VERSION 3.0.0)
+cmake_minimum_required(VERSION 3.0.0...3.21)
 
 list (APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/cmake")
 
@@ -6,8 +6,19 @@ include (DetectCPPZMQVersion)
 
 project(cppzmq VERSION ${DETECTED_CPPZMQ_VERSION})
 
-if (NOT TARGET libzmq AND NOT TARGET libzmq-static)
-  find_package(ZeroMQ QUIET)
+include(flags OPTIONAL)
+if(DEFINED XP_NAMESPACE)
+  set(nSpace ${XP_NAMESPACE}::)
+  set(nameSpace NAMESPACE ${nSpace})
+endif()
+
+if (NOT TARGET ${nSpace}libzmq AND NOT TARGET ${nSpace}libzmq-static)
+  if(COMMAND xpFindPkg)
+    xpFindPkg(PKGS libzmq)
+    set(ZeroMQ_FOUND TRUE)
+  else()
+    find_package(ZeroMQ QUIET)
+  endif()
 
   # libzmq autotools install: fallback to pkg-config
   if(NOT ZeroMQ_FOUND)
@@ -21,7 +32,7 @@ if (NOT TARGET libzmq AND NOT TARGET libzmq-static)
     message(FATAL_ERROR "ZeroMQ was not found, neither as a CMake package nor via pkg-config")
   endif()
 
-  if (ZeroMQ_FOUND AND NOT (TARGET libzmq OR TARGET libzmq-static))
+  if (ZeroMQ_FOUND AND NOT (TARGET ${nSpace}libzmq OR TARGET ${nSpace}libzmq-static))
     message(FATAL_ERROR "ZeroMQ version not supported!")
   endif()
 endif()
@@ -48,11 +59,11 @@ set(CPPZMQ_HEADERS
 foreach (target cppzmq cppzmq-static)
   add_library(${target} INTERFACE)
   target_include_directories(${target} INTERFACE $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}>
-                                                 $<INSTALL_INTERFACE:include>)
+                                                 $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>)
 endforeach()
 
-target_link_libraries(cppzmq INTERFACE libzmq)
-target_link_libraries(cppzmq-static INTERFACE libzmq-static)
+target_link_libraries(cppzmq INTERFACE ${nSpace}libzmq)
+target_link_libraries(cppzmq-static INTERFACE ${nSpace}libzmq-static)
 
 include(GNUInstallDirs)
 include(CMakePackageConfigHelpers)
@@ -64,7 +75,10 @@ install(FILES ${CPPZMQ_HEADERS}
         DESTINATION ${CMAKE_INSTALL_INCLUDEDIR})
 
 # GNUInstallDirs "DATADIR" wrong here; CMake search path wants "share".
-set(CPPZMQ_CMAKECONFIG_INSTALL_DIR "share/cmake/${PROJECT_NAME}" CACHE STRING "install path for cppzmqConfig.cmake")
+if(NOT DEFINED XP_INSTALL_CMAKEDIR)
+  set(XP_INSTALL_CMAKEDIR share/cmake/${PROJECT_NAME})
+endif()
+set(CPPZMQ_CMAKECONFIG_INSTALL_DIR "${XP_INSTALL_CMAKEDIR}" CACHE STRING "install path for cppzmqConfig.cmake")
 
 configure_file(libzmq-pkg-config/FindZeroMQ.cmake
                libzmq-pkg-config/FindZeroMQ.cmake
@@ -88,7 +102,7 @@ write_basic_package_version_file(${CMAKE_CURRENT_BINARY_DIR}/${PROJECT_NAME}Conf
 set(CMAKE_SIZEOF_VOID_P ${CPPZMQ_SIZEOF_VOID_P})
 
 install(EXPORT ${PROJECT_NAME}-targets
-        FILE ${PROJECT_NAME}Targets.cmake
+        FILE ${PROJECT_NAME}Targets.cmake ${nameSpace}
         DESTINATION ${CPPZMQ_CMAKECONFIG_INSTALL_DIR})
 install(FILES ${CMAKE_CURRENT_BINARY_DIR}/${PROJECT_NAME}Config.cmake
               ${CMAKE_CURRENT_BINARY_DIR}/${PROJECT_NAME}ConfigVersion.cmake
diff --git a/cppzmqConfig.cmake.in b/cppzmqConfig.cmake.in
index f596517..a24c2d1 100644
--- a/cppzmqConfig.cmake.in
+++ b/cppzmqConfig.cmake.in
@@ -4,8 +4,8 @@
 #
 # ::
 #
-#   cppzmq-static
-#   cppzmq
+#   @nSpace@cppzmq-static
+#   @nSpace@cppzmq
 #
 # This module sets the following variables in your project::
 #
@@ -31,6 +31,6 @@ endif()
 
 if(NOT TARGET @PROJECT_NAME@)
     include("${CMAKE_CURRENT_LIST_DIR}/@PROJECT_NAME@Targets.cmake")
-    get_target_property(@PROJECT_NAME@_INCLUDE_DIR cppzmq INTERFACE_INCLUDE_DIRECTORIES)
+    get_target_property(@PROJECT_NAME@_INCLUDE_DIR @nSpace@cppzmq INTERFACE_INCLUDE_DIRECTORIES)
 endif()
 
