diff --git a/CMakeLists.txt b/CMakeLists.txt
index 9ac3439..4633be1 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1,10 +1,18 @@
-cmake_minimum_required(VERSION 2.8.10)
+cmake_minimum_required(VERSION 3.0)
 project(ffmpeg)
 include(flags OPTIONAL)
+if(DEFINED FFMPEG_VER)
+  set(ver _${FFMPEG_VER})
+  set(verDir /${PROJECT_NAME}${ver})
+endif()
 if(${CMAKE_SYSTEM_NAME} STREQUAL "Windows")
-  file(GLOB lib ${CMAKE_CURRENT_SOURCE_DIR}/bin/*.lib)
-  file(GLOB dll ${CMAKE_CURRENT_SOURCE_DIR}/bin/*.dll)
-  install(FILES ${lib} DESTINATION lib${NUMBITS})
-  install(PROGRAMS ${dll} DESTINATION bin${NUMBITS})
+  file(GLOB libs ${CMAKE_CURRENT_SOURCE_DIR}/bin/*.lib)
+  file(GLOB dlls ${CMAKE_CURRENT_SOURCE_DIR}/bin/*.dll)
+  foreach(lib ${libs})
+    get_filename_component(ext ${lib} EXT)
+    get_filename_component(nam ${lib} NAME_WE)
+    install(FILES ${lib} DESTINATION lib RENAME ${nam}${ver}${ext})
+  endforeach()
+  install(PROGRAMS ${dlls} DESTINATION bin${verDir})
 endif()
-install(DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}/include/ DESTINATION include/${PROJECT_NAME})
+install(DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}/include/ DESTINATION include${verDir}/${PROJECT_NAME})
