diff --git a/CMakeLists.txt b/CMakeLists.txt
index 21e90151..97703567 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1,5 +1,5 @@
 # This is the legacy minimum version flatbuffers supported for a while.
-cmake_minimum_required(VERSION 2.8.12)
+cmake_minimum_required(VERSION 2.8.12...3.21)
 
 # CMake version 3.16 is the 'de-facto' minimum version for flatbuffers. If the
 # current cmake is older than this, warn the user and include the legacy file to
@@ -25,6 +25,7 @@ else()
   project(FlatBuffers)
 endif (POLICY CMP0048)
 
+include(flags OPTIONAL)
 include(CMake/Version.cmake)
 
 # generate compile_commands.json
@@ -427,6 +428,7 @@ if(FLATBUFFERS_BUILD_FLATC)
   if(FLATBUFFERS_STATIC_FLATC AND NOT MSVC)
     target_link_libraries(flatc PRIVATE -static)
   endif()
+  set_target_properties(flatc PROPERTIES DEBUG_POSTFIX "${CMAKE_DEBUG_POSTFIX}")
 endif()
 
 if(FLATBUFFERS_BUILD_FLATHASH)
@@ -647,11 +649,22 @@ if(FLATBUFFERS_INSTALL)
 
   install(DIRECTORY include/flatbuffers DESTINATION ${CMAKE_INSTALL_INCLUDEDIR})
 
-  set(FB_CMAKE_DIR "${CMAKE_INSTALL_LIBDIR}/cmake/flatbuffers")
+  if(DEFINED XP_NAMESPACE)
+    set(nameSpace ${XP_NAMESPACE}::)
+    set(XP_OPT_INSTALL EXCLUDE_FROM_ALL)
+  else()
+    set(nameSpace flatbuffers::)
+  endif()
+  if(DEFINED XP_INSTALL_CMAKEDIR)
+    set(FB_CMAKE_DIR ${XP_INSTALL_CMAKEDIR})
+  else()
+    set(FB_CMAKE_DIR "${CMAKE_INSTALL_DATADIR}/cmake/flatbuffers")
+  endif()
 
   configure_file(CMake/FlatbuffersConfigVersion.cmake.in FlatbuffersConfigVersion.cmake @ONLY)
   install(
       FILES "CMake/FlatbuffersConfig.cmake" "${CMAKE_CURRENT_BINARY_DIR}/FlatbuffersConfigVersion.cmake"
+        "CMake/BuildFlatBuffers.cmake" "CMake/FindFlatBuffers.cmake"
       DESTINATION ${FB_CMAKE_DIR}
   )
 
@@ -664,7 +677,7 @@ if(FLATBUFFERS_INSTALL)
 
     install(EXPORT FlatbuffersTargets
       FILE FlatbuffersTargets.cmake
-      NAMESPACE flatbuffers::
+      NAMESPACE ${nameSpace}
       DESTINATION ${FB_CMAKE_DIR}
     )
   endif()
@@ -673,12 +686,13 @@ if(FLATBUFFERS_INSTALL)
     install(
       TARGETS flatc EXPORT FlatcTargets
       RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR}
+      CONFIGURATIONS Release
     )
 
     install(
       EXPORT FlatcTargets
       FILE FlatcTargets.cmake
-      NAMESPACE flatbuffers::
+      NAMESPACE ${nameSpace}
       DESTINATION ${FB_CMAKE_DIR}
     )
   endif()
@@ -695,7 +709,7 @@ if(FLATBUFFERS_INSTALL)
     install(
       EXPORT FlatbuffersSharedTargets
       FILE FlatbuffersSharedTargets.cmake
-      NAMESPACE flatbuffers::
+      NAMESPACE ${nameSpace}
       DESTINATION ${FB_CMAKE_DIR}
     )
   endif()
@@ -704,6 +718,7 @@ if(FLATBUFFERS_INSTALL)
       configure_file(CMake/flatbuffers.pc.in flatbuffers.pc @ONLY)
       install(
         FILES "${CMAKE_CURRENT_BINARY_DIR}/flatbuffers.pc"
+        ${XP_OPT_INSTALL}
         DESTINATION ${CMAKE_INSTALL_LIBDIR}/pkgconfig
       )
   endif()
@@ -756,4 +771,4 @@ add_library(FlatBuffers::FlatBuffers ALIAS FlatBuffers)
 target_include_directories(
   FlatBuffers
   INTERFACE $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
-            $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}/include>)
\ No newline at end of file
+            $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}/include>)
