diff --git a/CMakeLists.txt b/CMakeLists.txt
index 3987eac9..bf194d11 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -4,6 +4,7 @@ set(CMAKE_EXPORT_COMPILE_COMMANDS ON)
 include(CheckCXXSymbolExists)
 
 project(FlatBuffers)
+include(flags OPTIONAL)
 
 # NOTE: Code coverage only works on Linux & OSX.
 option(FLATBUFFERS_CODE_COVERAGE "Enable the code coverage build option." OFF)
@@ -67,6 +68,18 @@ if(NOT DEFINED FLATBUFFERS_LOCALE_INDEPENDENT)
 endif()
 add_definitions(-DFLATBUFFERS_LOCALE_INDEPENDENT=$<BOOL:${FLATBUFFERS_LOCALE_INDEPENDENT}>)
 
+# version
+if(DEFINED FLATBUFFERS_VER)
+  set(_ver _${FLATBUFFERS_VER})
+  set(CMAKE_INSTALL_INCLUDEDIR include/flatbuffers${_ver})
+endif()
+# NAMESPACE
+if(DEFINED XP_NAMESPACE)
+  set(nameSpace ${XP_NAMESPACE}::)
+else()
+  set(nameSpace flatbuffers::)
+endif()
+
 set(FlatBuffers_Library_SRCS
   include/flatbuffers/base.h
   include/flatbuffers/flatbuffers.h
@@ -316,6 +329,7 @@ if(FLATBUFFERS_BUILD_FLATLIB)
   target_include_directories(flatbuffers INTERFACE
     $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>)
   target_compile_options(flatbuffers PRIVATE "${FLATBUFFERS_PRIVATE_CXX_FLAGS}")
+  set_target_properties(flatbuffers PROPERTIES OUTPUT_NAME flatbuffers${_ver})
 endif()
 
 if(FLATBUFFERS_BUILD_FLATC)
@@ -334,6 +348,9 @@ if(FLATBUFFERS_BUILD_FLATC)
   if(FLATBUFFERS_STATIC_FLATC AND NOT MSVC)
     target_link_libraries(flatc PRIVATE -static)
   endif()
+  set_target_properties(flatc PROPERTIES
+    DEBUG_POSTFIX "${CMAKE_DEBUG_POSTFIX}"
+    OUTPUT_NAME flatc${_ver})
 endif()
 
 if(FLATBUFFERS_BUILD_FLATHASH)
@@ -521,7 +538,7 @@ if(FLATBUFFERS_INSTALL)
 
   install(DIRECTORY include/flatbuffers DESTINATION ${CMAKE_INSTALL_INCLUDEDIR})
 
-  set(FB_CMAKE_DIR "${CMAKE_INSTALL_LIBDIR}/cmake/flatbuffers")
+  set(FB_CMAKE_DIR "${CMAKE_INSTALL_LIBDIR}/cmake/flatbuffers${_ver}")
 
   configure_file(CMake/FlatbuffersConfigVersion.cmake.in FlatbuffersConfigVersion.cmake @ONLY)
   install(
@@ -545,7 +562,7 @@ if(FLATBUFFERS_INSTALL)
 
     install(EXPORT FlatbuffersTargets
       FILE FlatbuffersTargets.cmake
-      NAMESPACE flatbuffers::
+      NAMESPACE ${nameSpace}
       DESTINATION ${FB_CMAKE_DIR}
     )
   endif()
@@ -559,7 +576,7 @@ if(FLATBUFFERS_INSTALL)
     install(
       EXPORT FlatcTargets
       FILE FlatcTargets.cmake
-      NAMESPACE flatbuffers::
+      NAMESPACE ${nameSpace}
       DESTINATION ${FB_CMAKE_DIR}
     )
   endif()
@@ -585,7 +602,7 @@ if(FLATBUFFERS_INSTALL)
   install(
       EXPORT FlatbuffersSharedTargets
       FILE FlatbuffersSharedTargets.cmake
-      NAMESPACE flatbuffers::
+      NAMESPACE ${nameSpace}
       DESTINATION ${FB_CMAKE_DIR}
     )
   endif()
