diff --git a/CMakeLists.txt b/CMakeLists.txt
new file mode 100644
index 0000000..2b3c540
--- /dev/null
+++ b/CMakeLists.txt
@@ -0,0 +1,4 @@
+cmake_minimum_required(VERSION 3.20)
+project(glew)
+include(flags OPTIONAL)
+add_subdirectory(build/cmake)
diff --git a/build/cmake/CMakeLists.txt b/build/cmake/CMakeLists.txt
index 1563f9f..feb0007 100644
--- a/build/cmake/CMakeLists.txt
+++ b/build/cmake/CMakeLists.txt
@@ -2,15 +2,21 @@ if ( NOT DEFINED CMAKE_BUILD_TYPE )
   set( CMAKE_BUILD_TYPE Release CACHE STRING "Build type" )
 endif ()
 
+if(CMAKE_SOURCE_DIR STREQUAL CMAKE_CURRENT_SOURCE_DIR)
 project (glew)
 
 cmake_minimum_required (VERSION 2.4)
+endif()
+
+include(GNUInstallDirs)
 
 if (COMMAND cmake_policy)
   cmake_policy (SET CMP0003 NEW)
 endif()
 
-set(CMAKE_DEBUG_POSTFIX d)
+if (NOT DEFINED CMAKE_DEBUG_POSTFIX)
+  set(CMAKE_DEBUG_POSTFIX d)
+endif ()
 
 option (BUILD_UTILS "utilities" ON)
 
@@ -64,7 +70,7 @@ else()
   target_compile_definitions(glewmx INTERFACE "GLEW_MX")
   target_compile_definitions(glewmx_s INTERFACE "GLEW_STATIC;GLEW_MX")
   foreach(t glew glew_s glewmx glewmx_s)
-    target_include_directories(${t} PUBLIC $<INSTALL_INTERFACE:include>)
+    target_include_directories(${t} PUBLIC $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>)
   endforeach()
   set(MAYBE_EXPORT EXPORT glew-targets)
 endif()
@@ -80,9 +86,9 @@ endif()
 
 install ( TARGETS ${targets_to_install}
           ${MAYBE_EXPORT}
-          RUNTIME DESTINATION bin
-          LIBRARY DESTINATION lib${LIB_SUFFIX}
-          ARCHIVE DESTINATION lib${LIB_SUFFIX}
+          RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR}
+          LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
+          ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
 )
 
 if (BUILD_UTILS)
@@ -93,7 +99,7 @@ if (BUILD_UTILS)
   target_link_libraries (visualinfo glew)
 
   install ( TARGETS glewinfo visualinfo
-            DESTINATION bin)
+            DESTINATION ${CMAKE_INSTALL_BINDDIR})
 endif ()
 
 set (prefix ${CMAKE_INSTALL_PREFIX})
@@ -110,21 +116,27 @@ set (cflags "-DGLEW_MX")
 set (libname ${GLEW_LIB_NAME}mx)
 configure_file (${GLEW_DIR}/glew.pc.in ${GLEW_DIR}/glewmx.pc @ONLY)
 
-install(FILES ${GLEW_DIR}/glew.pc ${GLEW_DIR}/glewmx.pc
-        DESTINATION lib/pkgconfig
-)
+option (INSTALL_PKGCONFIG "pkgconfig install" ON)
+if (INSTALL_PKGCONFIG)
+  install(FILES ${GLEW_DIR}/glew.pc ${GLEW_DIR}/glewmx.pc
+          DESTINATION ${CMAKE_INSTALL_DATADIR}/pkgconfig
+  )
+endif ()
 
 install (FILES
     ${GLEW_DIR}/include/GL/wglew.h
     ${GLEW_DIR}/include/GL/glew.h
     ${GLEW_DIR}/include/GL/glxew.h
-    DESTINATION include/GL)
+    DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/GL)
 
 if(MAYBE_EXPORT)
-  install(EXPORT glew-targets DESTINATION lib/cmake/glew
+  if(NOT DEFINED XP_INSTALL_CMAKEDIR)
+    set(XP_INSTALL_CMAKEDIR ${CMAKE_INSTALL_DATADIR}/cmake/glew)
+  endif()
+  install(EXPORT glew-targets DESTINATION ${XP_INSTALL_CMAKEDIR}
     NAMESPACE GLEW::)
   install(FILES
       ${CMAKE_CURRENT_SOURCE_DIR}/glew-config.cmake
       ${CMAKE_CURRENT_SOURCE_DIR}/CopyImportedTargetProperties.cmake
-    DESTINATION lib/cmake/glew)
+    DESTINATION ${XP_INSTALL_CMAKEDIR})
 endif()
