diff --git a/CMakeLists.txt b/CMakeLists.txt
index 3dccec310..54d3fcc89 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -32,6 +32,7 @@ INCLUDE(GNUInstallDirs)
 INCLUDE(IdeSplitSources)
 INCLUDE(FeatureSummary)
 INCLUDE(EnableWarnings)
+INCLUDE(flags OPTIONAL)
 
 # Build options
 #
@@ -58,6 +59,15 @@ OPTION(USE_BUNDLED_ZLIB    		"Use the bundled version of zlib. Can be set to one
 OPTION(DEPRECATE_HARD			"Do not include deprecated functions in the library"	OFF)
    SET(REGEX_BACKEND			"" CACHE STRING "Regular expression implementation. One of regcomp_l, pcre2, pcre, regcomp, or builtin.")
 
+OPTION(OPENSSL_MODULE_PATH		"Find OpenSSL in CMAKE_MODULE_PATH"			OFF)
+OPTION(ZLIB_MODULE_PATH			"Find zlib in CMAKE_MODULE_PATH"			OFF)
+OPTION(LIBSSH2_MODULE_PATH		"Find libssh2 in CMAKE_MODULE_PATH"			OFF)
+OPTION(INSTALL_LIBGIT2_CONFIG		"Install libgit2 config (pkg-config libgit2.pc)"	ON)
+MARK_AS_ADVANCED(OPENSSL_MODULE_PATH)
+MARK_AS_ADVANCED(ZLIB_MODULE_PATH)
+MARK_AS_ADVANCED(LIBSSH2_MODULE_PATH)
+MARK_AS_ADVANCED(INSTALL_LIBGIT2_CONFIG)
+
 IF (UNIX)
 	IF (NOT USE_HTTPS)
 	    OPTION(USE_NTLMCLIENT		"Enable NTLM support on Unix."				OFF )
@@ -79,6 +89,10 @@ IF(MSVC)
 	# are linking statically
 	OPTION(STATIC_CRT		"Link the static CRT libraries"		 		 ON)
 
+	# Some MSVC flags are not compatible with what's been set externally
+	OPTION(SKIP_MSVC_FLAGS		"Skip MSVC flags set here"				OFF)
+	MARK_AS_ADVANCED(SKIP_MSVC_FLAGS)
+
 	# If you want to embed a copy of libssh2 into libgit2, pass a
 	# path to libssh2
 	OPTION(EMBED_SSH_PATH		"Path to libssh2 to embed (Windows)"			OFF)
@@ -102,6 +116,8 @@ ENDIF()
 
 # Platform specific compilation flags
 IF (MSVC)
+	IF (NOT SKIP_MSVC_FLAGS)
+
 	IF (STDCALL)
 		MESSAGE(FATAL_ERROR "The STDCALL option is no longer supported; libgit2 is now always built as a cdecl library.  If you're using PInvoke, please add the CallingConventions.Cdecl attribute for support.")
 	ENDIF()
@@ -191,6 +207,8 @@ IF (MSVC)
 	SET(CMAKE_SHARED_LINKER_FLAGS_RELEASE "${CMAKE_EXE_LINKER_FLAGS_RELEASE}")
 	SET(CMAKE_SHARED_LINKER_FLAGS_RELWITHDEBINFO "${CMAKE_EXE_LINKER_FLAGS_RELWITHDEBINFO}")
 	SET(CMAKE_SHARED_LINKER_FLAGS_MINSIZEREL "${CMAKE_EXE_LINKER_FLAGS_MINSIZEREL}")
+
+	ENDIF() # SKIP_MSVC_FLAGS
 ELSE ()
 	IF (ENABLE_REPRODUCIBLE_BUILDS)
 		SET(CMAKE_C_ARCHIVE_CREATE "<CMAKE_AR> Dqc <TARGET> <LINK_FLAGS> <OBJECTS>")
diff --git a/cmake/PkgBuildConfig.cmake b/cmake/PkgBuildConfig.cmake
index 54c5e294c..bc8665e6e 100644
--- a/cmake/PkgBuildConfig.cmake
+++ b/cmake/PkgBuildConfig.cmake
@@ -72,6 +72,8 @@ function(pkg_build_config)
     endif()
     file(APPEND "${PKGCONFIG_FILE}" "Cflags: -I\${includedir} ${PKGCONFIG_CFLAGS}\n")
 
+    if(INSTALL_LIBGIT2_CONFIG)
     # Install .pc file
     install(FILES "${PKGCONFIG_FILE}" DESTINATION "${CMAKE_INSTALL_LIBDIR}/pkgconfig")
+    endif()
 endfunction()
diff --git a/cmake/SelectHTTPSBackend.cmake b/cmake/SelectHTTPSBackend.cmake
index 4998f0f2a..de60a38f8 100644
--- a/cmake/SelectHTTPSBackend.cmake
+++ b/cmake/SelectHTTPSBackend.cmake
@@ -1,7 +1,12 @@
 INCLUDE(SanitizeBool)
 
 # We try to find any packages our backends might use
-FIND_PACKAGE(OpenSSL)
+IF(OPENSSL_MODULE_PATH)
+	FIND_PACKAGE(usexp-OpenSSL REQUIRED PATHS ${CMAKE_MODULE_PATH} NO_DEFAULT_PATH)
+	GET_TARGET_PROPERTY(OPENSSL_INCLUDE_DIR xpro::crypto INTERFACE_INCLUDE_DIRECTORIES)
+ELSE()
+	FIND_PACKAGE(OpenSSL)
+ENDIF()
 FIND_PACKAGE(mbedTLS)
 IF (CMAKE_SYSTEM_NAME MATCHES "Darwin")
 	FIND_PACKAGE(Security)
diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index fdb367335..e2be8a394 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -35,6 +35,15 @@ SET(LIBGIT2_INCLUDES
 SET(LIBGIT2_SYSTEM_INCLUDES "")
 SET(LIBGIT2_LIBS "")
 
+IF(DEFINED LIBGIT2_VER)
+  SET(ver _${LIBGIT2_VER})
+  IF(DEFINED VER_MOD)
+    SET(ver ${ver}${VER_MOD})
+  ENDIF()
+  SET(LIBGIT2_FILENAME git2${ver})
+ENDIF()
+SET(LIBGIT2_TARGETS_FILE ${PROJECT_NAME}${ver}-targets)
+
 enable_warnings(missing-declarations)
 
 # Enable tracing
@@ -203,7 +212,12 @@ IF(USE_BUNDLED_ZLIB STREQUAL ON)
 ENDIF()
 
 IF(USE_BUNDLED_ZLIB STREQUAL "OFF")
-	FIND_PACKAGE(ZLIB)
+	IF(ZLIB_MODULE_PATH)
+		FIND_PACKAGE(usexp-ZLIB PATHS ${CMAKE_MODULE_PATH} NO_DEFAULT_PATH)
+		GET_TARGET_PROPERTY(ZLIB_INCLUDE_DIRS ${ZLIB_LIBRARIES} INTERFACE_INCLUDE_DIRECTORIES)
+	ELSE()
+		FIND_PACKAGE(ZLIB)
+	ENDIF()
 	IF(ZLIB_FOUND)
 		LIST(APPEND LIBGIT2_SYSTEM_INCLUDES ${ZLIB_INCLUDE_DIRS})
 		LIST(APPEND LIBGIT2_LIBS ${ZLIB_LIBRARIES})
@@ -231,7 +245,12 @@ ENDIF()
 
 # Optional external dependency: libssh2
 IF (USE_SSH)
-	FIND_PKGLIBRARIES(LIBSSH2 libssh2)
+	IF(LIBSSH2_MODULE_PATH)
+		FIND_PACKAGE(usexp-libssh2 PATHS ${CMAKE_MODULE_PATH} NO_DEFAULT_PATH)
+		GET_TARGET_PROPERTY(LIBSSH2_INCLUDE_DIRS ${LIBSSH2_LIBRARIES} INTERFACE_INCLUDE_DIRECTORIES)
+	ELSE()
+		FIND_PKGLIBRARIES(LIBSSH2 libssh2)
+	ENDIF()
 	IF (NOT LIBSSH2_FOUND)
 		FIND_PACKAGE(LibSSH2)
 		SET(LIBSSH2_INCLUDE_DIRS ${LIBSSH2_INCLUDE_DIR})
@@ -384,6 +403,7 @@ ENDIF()
 # Compile and link libgit2
 ADD_LIBRARY(git2 ${WIN_RC} ${LIBGIT2_OBJECTS})
 TARGET_LINK_LIBRARIES(git2 ${LIBGIT2_LIBS})
+TARGET_INCLUDE_DIRECTORIES(git2 PUBLIC $<INSTALL_INTERFACE:$<INSTALL_PREFIX>/${CMAKE_INSTALL_INCLUDEDIR}>)
 
 SET_TARGET_PROPERTIES(git2 PROPERTIES C_STANDARD 90)
 SET_TARGET_PROPERTIES(git2 PROPERTIES LIBRARY_OUTPUT_DIRECTORY ${libgit2_BINARY_DIR})
@@ -424,10 +444,14 @@ IF (MSVC_IDE)
 ENDIF ()
 
 # Install
-INSTALL(TARGETS git2
+INSTALL(TARGETS git2 EXPORT ${LIBGIT2_TARGETS_FILE}
 	RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR}
 	LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
 	ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
 )
 INSTALL(DIRECTORY ${libgit2_SOURCE_DIR}/include/git2 DESTINATION ${CMAKE_INSTALL_INCLUDEDIR})
 INSTALL(FILES ${libgit2_SOURCE_DIR}/include/git2.h DESTINATION ${CMAKE_INSTALL_INCLUDEDIR})
+IF(DEFINED XP_NAMESPACE)
+  SET(nameSpace NAMESPACE ${XP_NAMESPACE}::)
+ENDIF()
+INSTALL(EXPORT ${LIBGIT2_TARGETS_FILE} DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake ${nameSpace})
