diff --git a/CMakeLists.txt b/CMakeLists.txt
new file mode 100644
index 0000000..8b96b1d
--- /dev/null
+++ b/CMakeLists.txt
@@ -0,0 +1,32 @@
+cmake_minimum_required(VERSION 3.24)
+project(node-addon-api)
+include(GNUInstallDirs)
+include(flags OPTIONAL)
+if(COMMAND xpFindPkg)
+  xpFindPkg(PKGS node)
+endif()
+set(lib_name ${PROJECT_NAME})
+########################################
+set(libsrcs
+  napi-inl.deprecated.h
+  napi-inl.h
+  napi.h
+  )
+########################################
+add_library(${lib_name} INTERFACE ${libsrcs})
+target_include_directories(${lib_name} INTERFACE $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>)
+target_compile_definitions(${lib_name} INTERFACE NODE_ADDON_API_DISABLE_DEPRECATED NAPI_CPP_EXCEPTIONS)
+if(TARGET xpro::node)
+  target_link_libraries(${lib_name} INTERFACE xpro::node)
+endif()
+########################################
+set(targetsFile ${PROJECT_NAME}-targets)
+install(TARGETS ${lib_name} EXPORT ${targetsFile})
+install(FILES ${libsrcs} DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/${PROJECT_NAME})
+if(DEFINED XP_NAMESPACE)
+  set(nameSpace NAMESPACE ${XP_NAMESPACE}::)
+endif()
+if(NOT DEFINED XP_INSTALL_CMAKEDIR)
+  set(XP_INSTALL_CMAKEDIR ${CMAKE_INSTALL_DATADIR}/cmake)
+endif()
+install(EXPORT ${targetsFile} DESTINATION ${XP_INSTALL_CMAKEDIR} ${nameSpace})
