diff --git a/openh264.pc.in b/openh264.pc.in
index 7fb5d0c..1a6b39b 100644
--- a/openh264.pc.in
+++ b/openh264.pc.in
@@ -1,10 +1,10 @@
 prefix=@prefix@
 libdir=${prefix}/lib
-includedir=${prefix}/include
+includedir=${prefix}/include@verDir@
 
 Name: OpenH264
 Description: OpenH264 is a codec library which supports H.264 encoding and decoding. It is suitable for use in real time applications such as WebRTC.
 Version: @VERSION@
-Libs: -L${libdir} -lopenh264 @LIBS@
+Libs: -L${libdir} -lopenh264@SUFFIX@ @LIBS@
 Libs.private: @LIBS_PRIVATE@
 Cflags: -I${includedir}
