diff --git a/CMakeLists.txt b/CMakeLists.txt
index 4a5bba6..d9fe257 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1,11 +1,22 @@
 cmake_minimum_required(VERSION 3.8)
 project(SQLite3
-    VERSION   3.31.1
+    VERSION   3.37.2
     LANGUAGES C
     )
 
 include(GNUInstallDirs)
 
+#------------------------------------------------------------------------------
+# externpro
+include(flags OPTIONAL)
+if(NOT DEFINED XP_NAMESPACE)
+  set(XP_NAMESPACE SQLite)
+  set(debugPostfix DEBUG_POSTFIX d)
+endif()
+if(NOT DEFINED XP_INSTALL_CMAKEDIR)
+  set(XP_INSTALL_CMAKEDIR cmake)
+endif()
+
 #------------------------------------------------------------------------------
 # build options and optional modules:
 option(SQLITE_ENABLE_DBSTAT_VTAB    "enables dbstat virtual table"                    OFF)
@@ -33,9 +44,9 @@ add_library(${PROJECT_NAME} STATIC sqlite3.c)
 set_target_properties(${PROJECT_NAME} PROPERTIES
     OUTPUT_NAME   sqlite3
     PUBLIC_HEADER sqlite3.h
-    DEBUG_POSTFIX d
+    ${debugPostfix}
     )
-target_include_directories(${PROJECT_NAME} PUBLIC $<INSTALL_INTERFACE:include>)
+target_include_directories(${PROJECT_NAME} PUBLIC $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>)
 target_compile_definitions(${PROJECT_NAME} PUBLIC # inject user's options
     $<BUILD_INTERFACE:
         $<$<BOOL:${SQLITE_ENABLE_DBSTAT_VTAB}>:SQLITE_ENABLE_DBSTAT_VTAB>
@@ -79,6 +90,13 @@ elseif(WIN32 AND ${CMAKE_SIZEOF_VOID_P} LESS 8) # this is a 32bit windows
             )
     endif()
 endif()
+if(SQLITE_ENABLE_MATH_FUNCTIONS)
+  include(CheckLibraryExists)
+  check_library_exists(m sin "" HAVE_LIB_M)
+  if(HAVE_LIB_M)
+    target_link_libraries(${PROJECT_NAME} PRIVATE m)
+  endif()
+endif()
 
 #------------------------------------------------------------------------------
 configure_file(sqlite3_config.h.in ${CMAKE_BINARY_DIR}/sqlite3_config.h)
@@ -89,8 +107,8 @@ install(TARGETS ${PROJECT_NAME} EXPORT ${PROJECT_NAME}Config
     PUBLIC_HEADER DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}
     )
 install(EXPORT  ${PROJECT_NAME}Config
-    NAMESPACE   SQLite::
-    DESTINATION cmake
+    NAMESPACE   ${XP_NAMESPACE}::
+    DESTINATION ${XP_INSTALL_CMAKEDIR}
     )
 install(FILES
     ${CMAKE_BINARY_DIR}/sqlite3_config.h DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}
@@ -120,7 +138,7 @@ if(BUILD_SHELL)
             set(${flag} "${${flag}}" CACHE STRING "msvc flags" FORCE)
         endforeach()
     endif()
-    install(TARGETS shell_app
-        RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR}
+    install(TARGETS shell_app EXPORT ${PROJECT_NAME}Config
+        RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR} CONFIGURATIONS Release
         )
 endif()
