diff --git a/CMakeLists.txt b/CMakeLists.txt
index f60caa1..5403966 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -52,14 +52,6 @@ set (WEBSOCKETPP_BUILD_ROOT ${CMAKE_CURRENT_BINARY_DIR})
 set (WEBSOCKETPP_BIN ${WEBSOCKETPP_BUILD_ROOT}/bin)
 set (WEBSOCKETPP_LIB ${WEBSOCKETPP_BUILD_ROOT}/lib)
 
-# CMake install step prefix. I assume linux users want the prefix to
-# be the default /usr or /usr/local so this is only adjusted on Windows.
-# - Windows: Build the INSTALL project in your solution file.
-# - Linux/OSX: make install.
-if (MSVC)
-    set (CMAKE_INSTALL_PREFIX "${WEBSOCKETPP_ROOT}/install")
-endif ()
-
 ############  Build customization
 
 # Override from command line "CMake -D<OPTION>=TRUE/FALSE/0/1/ON/OFF"
diff --git a/cmake/CMakeHelpers.cmake b/cmake/CMakeHelpers.cmake
index c2a20e5..ee64cbe 100644
--- a/cmake/CMakeHelpers.cmake
+++ b/cmake/CMakeHelpers.cmake
@@ -83,7 +83,7 @@ macro (final_target)
     file (GLOB_RECURSE TARGET_INSTALL_HEADERS RELATIVE ${CMAKE_CURRENT_SOURCE_DIR} *.hpp)
     foreach (hppfile ${TARGET_INSTALL_HEADERS})
       get_filename_component (currdir ${hppfile} PATH)
-      install (FILES ${hppfile} DESTINATION "include/${TARGET_NAME}/${currdir}")
+      install (FILES ${hppfile} DESTINATION "include${verDir}/${TARGET_NAME}/${currdir}")
     endforeach()
 endmacro ()
 
