diff --git a/CMakeLists.txt b/CMakeLists.txt
new file mode 100644
index 0000000..b8a86b4
--- /dev/null
+++ b/CMakeLists.txt
@@ -0,0 +1,20 @@
+cmake_minimum_required(VERSION 3.20)
+project(wxInclude)
+include(GNUInstallDirs)
+include(flags)
+set(XP_MODULE_PATH ${CMAKE_MODULE_PATH})
+xpFindPkg(PKGS boost)
+add_executable(${CMAKE_PROJECT_NAME} wxInclude.cpp)
+target_link_libraries(${CMAKE_PROJECT_NAME} PRIVATE Boost::filesystem Boost::program_options Boost::timer)
+string(TOLOWER ${PROJECT_NAME} prj)
+install(TARGETS ${CMAKE_PROJECT_NAME} EXPORT ${prj}-targets
+  RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR}
+  LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
+  ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
+  )
+if(DEFINED XP_NAMESPACE)
+  set(nameSpace NAMESPACE ${XP_NAMESPACE}::)
+endif()
+install(EXPORT ${prj}-targets DESTINATION ${CMAKE_INSTALL_BINDIR}/cmake ${nameSpace})
+add_executable(decompressor decompress.cpp)
+target_link_libraries(decompressor PRIVATE Boost::filesystem Boost::iostreams)
diff --git a/decompress.cpp b/decompress.cpp
new file mode 100644
index 0000000..d3aeb53
--- /dev/null
+++ b/decompress.cpp
@@ -0,0 +1,73 @@
+#include <fstream>
+#include <iostream>
+#include <boost/filesystem.hpp>
+#include <boost/iostreams/filtering_streambuf.hpp>
+#pragma warning(push)
+// conversion from std::streamsize to int, possible loss of data
+#pragma warning(disable : 4244)
+#include <boost/iostreams/copy.hpp>
+#pragma warning(pop)
+#include <boost/iostreams/filter/bzip2.hpp>
+#include <boost/iostreams/filter/gzip.hpp>
+#include <boost/iostreams/filter/zlib.hpp>
+
+int main(int argc, char** argv)
+{
+  namespace bfs = boost::filesystem;
+  namespace bio = boost::iostreams;
+  bfs::path exepath = bfs::path(std::string(argv[0]));
+  if (argc != 2)
+  {
+    std::cerr << "usage: " << exepath.filename().string() << " file.[bz2|gz|Z]" << std::endl;
+    return 1;
+  }
+  try
+  {
+    bfs::path filepath = bfs::path(std::string(argv[1]));
+    if (!bfs::exists(filepath))
+    {
+      std::cerr << filepath.string() << ": doesn't exist" << std::endl;
+      return 1;
+    }
+    bio::filtering_streambuf<bio::input> in;
+    std::string ext = filepath.extension().string();
+    if (ext.compare(".bz2") == 0)
+      in.push(bio::bzip2_decompressor());
+    else if (ext.compare(".gz") == 0)
+      in.push(bio::gzip_decompressor());
+    else if (ext.compare(".Z") == 0)
+      in.push(bio::zlib_decompressor());
+    else
+    {
+      std::cerr << filepath.filename().string() << ": unsupported extension (must be .[bz2|gz|Z])" << std::endl;
+      return 1;
+    }
+    std::ifstream file(argv[1], std::ios_base::in | std::ios_base::binary);
+    in.push(file);
+    bio::copy(in, std::cout);
+  }
+  //catch (const bio::bzip2_error& e)
+  catch (const bio::zlib_error& e)
+  {
+    int err = e.error();
+    if (err == bio::zlib::buf_error)
+      std::cerr << "zlib buffer error" << std::endl;
+    else if (err == bio::zlib::data_error)
+      std::cerr << "zlib data error" << std::endl;
+    else if (err == bio::zlib::mem_error)
+      std::cerr << "zlib memory error" << std::endl;
+    else if (err == bio::zlib::stream_error)
+      std::cerr << "zlib stream error" << std::endl;
+    else if (err == bio::zlib::version_error)
+      std::cerr << "zlib version error" << std::endl;
+    else
+      std::cerr << "zlib unknown error" << std::endl;
+    return 1;
+  }
+  catch (const std::exception& e)
+  {
+    std::cerr << e.what() << std::endl;
+    return 1;
+  }
+  return 0;
+}
diff --git a/wxInclude.cpp b/wxInclude.cpp
index c2109e3..04b8d1b 100644
--- a/wxInclude.cpp
+++ b/wxInclude.cpp
@@ -1,20 +1,25 @@
 /*
-  wxInclude version 1.0
+  wxInclude
   Kim De Deyn
 */
 
 #include <iostream>
 #include <boost/algorithm/string.hpp>
+#include <boost/chrono.hpp>
 #include <boost/filesystem.hpp>
 #include <boost/filesystem/fstream.hpp>
 #include <boost/foreach.hpp>
 #include <boost/format.hpp>
 #include <boost/program_options.hpp>
-#include <boost/timer.hpp>
+#include <boost/timer/timer.hpp>
 
 namespace po = boost::program_options;
 namespace fs = boost::filesystem;
 
+#ifdef _MSC_VER
+#define snprintf _snprintf
+#endif
+
 #define WXINCLUDE_INFO	"wxInclude by Kim De Deyn, use --help for more information.\n"
 
 #define WXINCLUDE_HELP	"This tool can be used to convert binary files into a useful C/C++ header.\n" \
@@ -25,9 +30,11 @@ namespace fs = boost::filesystem;
   "Example of use:\n\n" \
   "  wxInclude.exe --const --input-file=mydata1.bin\n" \
   "    --input-type=.png --input-type=.bmp\n" \
-  "    --output-file=myheader.h mydata2.bin myimage.png\n"
+  "    --output-file=myheader.h mydata2.bin myimage.png\n\n" \
+  "Recommended usage (at Space Dynamics Lab):\n" \
+  "  wxInclude --const --appendtype --wxnone --respectcase --output-file=foo.hrc foo.png\n"
 
-#define WXINCLUDE_VERSION "Version 1.0, compiled at " __DATE__ " " __TIME__
+#define WXINCLUDE_VERSION "Version 1.2, compiled on " __DATE__ " at " __TIME__
 
 #define BUFFER_SIZE 4096
 
@@ -106,7 +113,7 @@ void definefile( std::ostringstream& data, fs::ifstream& input, std::string& nam
   /* Define array */
   data << "static" << ( useconst ? " const " : " " ) << "unsigned char " << name << "[] = {" << std::endl;
 
-  int size = input.tellg();
+  int size = static_cast<int>(input.tellg()); // conversion from std::streamoff to int, possible loss of data
   input.seekg( 0, std::ios::beg );
 
   int c = 0;
@@ -130,7 +137,7 @@ void definefile( std::ostringstream& data, fs::ifstream& input, std::string& nam
       Still faster then the optimized boost::format use, but not that much!
     */
     static char temp[5];
-    _snprintf( temp, 5, "0x%02X", c );
+    snprintf( temp, 5, "0x%02X", c );
     data << temp;
 
     if ( i >= size )
@@ -156,7 +163,7 @@ void definefile( std::ostringstream& data, fs::ifstream& input, std::string& nam
   data << "};" << std::endl << std::endl;
 }
 
-int _tmain(int argc, _TCHAR* argv[])
+int main(int argc, char* argv[])
 {
   try
   {
@@ -164,9 +171,10 @@ int _tmain(int argc, _TCHAR* argv[])
     desc.add_options()
       ( "help,h", "Show detailed help." )
       ( "options,p", "Show parameter information." )
-      ( "version,p", "Show version information." )
-      ( "input-file,i", po::value<std::vector<std::string>>(), "Define file(s) for the convertion input." )
-      ( "input-type,I", po::value<std::vector<std::string>>(), "Define file type(s) for automatic conversion of files in the working directory." )
+      ( "version,v", "Show version information." )
+      ( "quiet,q", "Quiet at runtime, not verbose." )
+      ( "input-file,i", po::value<std::vector<std::string> >(), "Define file(s) for the convertion input." )
+      ( "input-type,I", po::value<std::vector<std::string> >(), "Define file type(s) for automatic conversion of files in the working directory." )
       ( "output-file,o", po::value<std::string>(), "Define file for the convertion output." )
       ( "noheader,h", "Disable adding of header support defines." )
       ( "const,C", "Define array as const." )
@@ -174,7 +182,6 @@ int _tmain(int argc, _TCHAR* argv[])
       ( "wxnone,w", "Disable adding of wxWidgets support macro's." )
       ( "wxheader,W", po::value<std::string>()->default_value( "wx/wx.h" ), "Select the header that includes wxWidgets (precompiled header?)." )
       ( "appendtype,t", "Add the file type at the end of the identifier (myimage_png)." )
-                        ( "text,T", "Disable binary output and use text output, converts feed codes to systems defaults." )
     ;
 
     po::positional_options_description posdesc;
@@ -182,10 +189,12 @@ int _tmain(int argc, _TCHAR* argv[])
 
     po::variables_map opt;
     po::store( po::command_line_parser( argc, argv ).options( desc ).positional( posdesc ).run(), opt );
-    po::store( po::parse_config_file( fs::ifstream( fs::path( "default.cfg" ) ), desc ), opt );
+    fs::ifstream ifs(fs::path("default.cfg"));
+    po::store( po::parse_config_file( ifs, desc ), opt );
     po::notify( opt );
 
-    std::cout << WXINCLUDE_INFO << std::endl;
+    if ( !opt.count( "quiet" ) )
+      std::cout << WXINCLUDE_INFO << std::endl;
 
     /* Show options when requested */
     if ( opt.count( "options" ) )
@@ -215,26 +224,23 @@ int _tmain(int argc, _TCHAR* argv[])
       if ( opt.count( "output-file" ) )
       {
         /* Create timer */
-        boost::timer timer;
+        boost::timer::cpu_timer timer;
 
         /* Create output file */
         std::string headername( opt[ "output-file" ].as<std::string>() );
 
         fs::path outputpath( headername );
-        fs::ofstream output( outputpath, std::ios::out | std::ios::trunc );
+        fs::ofstream output( outputpath, std::ios::out | std::ios::trunc | std::ios::binary );
 
         /* Use buffer */
         char outbuffer[BUFFER_SIZE];
         output.rdbuf()->pubsetbuf( outbuffer, BUFFER_SIZE );
 
-                                if ( !opt.count( "text" ) )
-                                        output.setf( std::ios::binary );
-
         if ( !output )
           throw std::runtime_error( "Failed to create output file!" );
 
-        /* Show status */
-        std::cout << "Build  : file '" << outputpath.leaf() << "'..." << std::endl;
+        if ( !opt.count( "quiet" ) ) /* Show status */
+          std::cout << "Build  : file '" << outputpath.leaf() << "'..." << std::endl;
 
         /* Get base name of file */
         headername = fs::basename( outputpath );
@@ -259,7 +265,7 @@ int _tmain(int argc, _TCHAR* argv[])
         /* Process input files based on provided list */
         if ( opt.count( "input-file" ) )
         {
-          std::vector<std::string> files( opt[ "input-file" ].as<std::vector<std::string>>() );
+          std::vector<std::string> files( opt[ "input-file" ].as<std::vector<std::string> >() );
 
           BOOST_FOREACH( std::string& file, files )
           {
@@ -271,8 +277,8 @@ int _tmain(int argc, _TCHAR* argv[])
 
             if ( input.is_open() )
             {
-              /* Show status */
-              std::cout << "Process: file '" << file << "'..." << std::endl;
+              if ( !opt.count( "quiet" ) ) /* Show status */
+                std::cout << "Process: file '" << file << "'..." << std::endl;
 
               /* Remove extension */
               boost::erase_last( file, fileext );
@@ -308,7 +314,7 @@ int _tmain(int argc, _TCHAR* argv[])
         /* Process input files based on provided type */
         if ( opt.count( "input-type" ) )
         {
-          std::vector<std::string> types( opt[ "input-type" ].as<std::vector<std::string>>() );
+          std::vector<std::string> types( opt[ "input-type" ].as<std::vector<std::string> >() );
 
           for ( fs::directory_iterator dir_itr( fs::initial_path() ); dir_itr != fs::directory_iterator(); ++dir_itr )
           {
@@ -332,12 +338,12 @@ int _tmain(int argc, _TCHAR* argv[])
                   fs::ifstream input( dir_itr->path(), std::ios::in | std::ios::binary | std::ios::ate );
                   input.rdbuf()->pubsetbuf( inbuffer, BUFFER_SIZE );
 
-                  std::string file( dir_itr->path().leaf() );
+                  std::string file( dir_itr->path().leaf().string() );
 
                   if ( input.is_open() )
                   {
-                    /* Show status */
-                    std::cout << "Process: file '" << file << "'..." << std::endl;
+                    if ( !opt.count( "quiet" ) ) /* Show status */
+                      std::cout << "Process: file '" << file << "'..." << std::endl;
 
                     /* Remove extension */
                     boost::erase_last( file, fileext );
@@ -378,8 +384,12 @@ int _tmain(int argc, _TCHAR* argv[])
         output.seekp( 0, std::ios::beg );
         output << data.str();
 
-        /* Show status */
-        std::cout << "Build  : " << timer.elapsed() << "s needed for conversion of " << list.size() << " files." << std::endl;
+        if ( !opt.count( "quiet" ) ) /* Show status */
+        {
+          auto nanoseconds = boost::chrono::nanoseconds(timer.elapsed().user + timer.elapsed().system);
+          auto seconds = boost::chrono::duration_cast<boost::chrono::seconds>(nanoseconds);
+          std::cout << "Build  : " << seconds.count() << "s needed for conversion of " << list.size() << " files." << std::endl;
+        }
       }
       else
       {
