diff --git a/CMakeLists.txt b/CMakeLists.txt
new file mode 100644
index 0000000..145f178
--- /dev/null
+++ b/CMakeLists.txt
@@ -0,0 +1,31 @@
+set(lib_name tlc) # wxTreeListCtrl
+#######################################
+# include directories
+include_directories(include)
+#######################################
+# library sources
+########################
+# source
+set(treelist_srcs
+  src/treelistctrl.cpp
+  )
+list(APPEND ${lib_name}_libsrcs ${treelist_srcs})
+########################
+# headers
+set(treelist_hdrs
+  include/wx/treelistctrl.h
+  )
+list(APPEND ${lib_name}_libsrcs ${treelist_hdrs})
+#######################################
+# library
+add_library(${lib_name} STATIC ${${lib_name}_libsrcs})
+set_wxxtarget_properties(${lib_name})
+#######################################
+# install
+install(TARGETS ${lib_name} EXPORT ${PROJECT_NAME}${WX_VER}-targets
+  RUNTIME DESTINATION bin
+  LIBRARY DESTINATION lib
+  ARCHIVE DESTINATION lib
+  )
+install(FILES ${treelist_hdrs} DESTINATION include/${wxinc}/wx/treelistctrl)
+install(EXPORT ${PROJECT_NAME}${WX_VER}-targets DESTINATION lib/cmake)
diff --git a/src/treelistctrl.cpp b/src/treelistctrl.cpp
index 4189e37..10def35 100644
--- a/src/treelistctrl.cpp
+++ b/src/treelistctrl.cpp
@@ -50,10 +50,6 @@
 #include <wx/dynarray.h>
 #include <wx/arrimpl.cpp>
 
-#ifdef __WXMAC__
-#include "wx/mac/private.h"
-#endif
-
 #include "wx/treelistctrl.h"
 
 #include <wx/log.h>  // only required for debugging purpose
@@ -1289,7 +1285,7 @@ void wxTreeListHeaderWindow::DoDrawRect( wxDC *dc, int x, int y, int w, int h )
 #if !wxCHECK_VERSION(2, 5, 0)
     wxPen pen (wxSystemSettings::GetSystemColour (wxSYS_COLOUR_BTNSHADOW ), 1, wxSOLID);
 #else
-    wxPen pen (wxSystemSettings::GetColour (wxSYS_COLOUR_BTNSHADOW ), 1, wxSOLID);
+    wxPen pen (wxSystemSettings::GetColour (wxSYS_COLOUR_BTNSHADOW ), 1, wxPENSTYLE_SOLID);
 #endif
 
     const int m_corner = 1;
@@ -1487,7 +1483,7 @@ void wxTreeListHeaderWindow::DrawCurrent()
 
     wxScreenDC dc;
     dc.SetLogicalFunction (wxINVERT);
-    dc.SetPen (wxPen (*wxBLACK, 2, wxSOLID));
+    dc.SetPen (wxPen (*wxBLACK, 2, wxPENSTYLE_SOLID));
     dc.SetBrush (*wxTRANSPARENT_BRUSH);
 
     AdjustDC(dc);
@@ -1969,8 +1965,8 @@ void wxTreeListMainWindow::Init() {
     m_hilightBrush = new wxBrush (wxSystemSettings::GetSystemColour (wxSYS_COLOUR_HIGHLIGHT), wxSOLID);
     m_hilightUnfocusedBrush = new wxBrush (wxSystemSettings::GetSystemColour (wxSYS_COLOUR_BTNSHADOW), wxSOLID);
 #else
-    m_hilightBrush = new wxBrush (wxSystemSettings::GetColour (wxSYS_COLOUR_HIGHLIGHT), wxSOLID);
-    m_hilightUnfocusedBrush = new wxBrush (wxSystemSettings::GetColour (wxSYS_COLOUR_BTNSHADOW), wxSOLID);
+    m_hilightBrush = new wxBrush (wxSystemSettings::GetColour (wxSYS_COLOUR_HIGHLIGHT), wxBRUSHSTYLE_SOLID);
+    m_hilightUnfocusedBrush = new wxBrush (wxSystemSettings::GetColour (wxSYS_COLOUR_BTNSHADOW), wxBRUSHSTYLE_SOLID);
 #endif
 
     m_imageListNormal = (wxImageList *) NULL;
@@ -2004,7 +2000,7 @@ void wxTreeListMainWindow::Init() {
     m_boldFont = wxFont( m_normalFont.GetPointSize(),
                          m_normalFont.GetFamily(),
                          m_normalFont.GetStyle(),
-                         wxBOLD,
+                         wxFONTWEIGHT_BOLD,
                          m_normalFont.GetUnderlined(),
                          m_normalFont.GetFaceName(),
                          m_normalFont.GetEncoding());
@@ -2023,8 +2019,6 @@ bool wxTreeListMainWindow::Create (wxTreeListCtrl *parent,
                                    const wxString& name) {
 
 #ifdef __WXMAC__
-    if (style & wxTR_HAS_BUTTONS) style |= wxTR_MAC_BUTTONS;
-    if (style & wxTR_HAS_BUTTONS) style &= ~wxTR_HAS_BUTTONS;
     style &= ~wxTR_LINES_AT_ROOT;
     style |= wxTR_NO_LINES;
 
@@ -2067,7 +2061,7 @@ bool wxTreeListMainWindow::Create (wxTreeListCtrl *parent,
     }
 #else
 //?    m_dottedPen = wxPen( *wxGREY_PEN, 1, wxDOT );  // too slow under XFree86
-    m_dottedPen = wxPen( _T("grey"), 0, 0 ); // Bitmap based pen is not supported by GTK!
+    m_dottedPen = wxPen( _T("grey"), 0, wxPENSTYLE_DOT ); // Bitmap based pen is not supported by GTK!
 #endif
 
     m_owner = parent;
@@ -2334,7 +2328,7 @@ bool wxTreeListMainWindow::SetFont (const wxFont &font) {
     m_boldFont = wxFont (m_normalFont.GetPointSize(),
                          m_normalFont.GetFamily(),
                          m_normalFont.GetStyle(),
-                         wxBOLD,
+                         wxFONTWEIGHT_BOLD,
                          m_normalFont.GetUnderlined(),
                          m_normalFont.GetFaceName());
     CalculateLineHeight();
@@ -3392,7 +3386,7 @@ void wxTreeListMainWindow::PaintItem (wxTreeListItem *item, wxDC& dc) {
 
 // determine background and show it
 // in wxTR_FULL_ROW_HIGHLIGHT mode, some drawing can be done already now
-    dc.SetBrush (wxBrush ( colBg, wxSOLID));
+    dc.SetBrush (wxBrush ( colBg, wxBRUSHSTYLE_SOLID));
     dc.SetPen (*wxTRANSPARENT_PEN);
     if (HasFlag (wxTR_FULL_ROW_HIGHLIGHT)) {
         if (item->IsSelected()) {
@@ -3482,7 +3476,7 @@ void wxTreeListMainWindow::PaintItem (wxTreeListItem *item, wxDC& dc) {
             if (item->IsSelected() && i == GetMainColumn()) {
                 // draw normal background
                 dc.SetPen (*wxTRANSPARENT_PEN);
-                dc.SetBrush (wxBrush ( colBg, wxSOLID));
+                dc.SetBrush (wxBrush ( colBg, wxBRUSHSTYLE_SOLID));
                 dc.DrawRectangle (x_colstart, item->GetY() + off_h, col_w, total_h - off_h);
                 // draw selection & optionally cursor
                 dc.SetPen (drawCursor ? *wxBLACK_PEN : *wxTRANSPARENT_PEN);
@@ -3493,7 +3487,7 @@ void wxTreeListMainWindow::PaintItem (wxTreeListItem *item, wxDC& dc) {
             } else {
                 // draw normal background & optionally cursor
                 dc.SetPen (drawCursor && i != GetMainColumn() ? *wxBLACK_PEN : *wxTRANSPARENT_PEN);
-                dc.SetBrush (wxBrush ( colBg, wxSOLID));
+                dc.SetBrush (wxBrush ( colBg, wxBRUSHSTYLE_SOLID));
                 dc.SetTextForeground (colText);
                 dc.DrawRectangle (x_colstart, item->GetY() + off_h, col_w, total_h - off_h);
                 // on main col draw a separate cursor
@@ -3510,7 +3504,7 @@ void wxTreeListMainWindow::PaintItem (wxTreeListItem *item, wxDC& dc) {
 #if !wxCHECK_VERSION(2, 5, 0)
             wxPen pen (wxSystemSettings::GetSystemColour (wxSYS_COLOUR_3DLIGHT ), 1, wxSOLID);
 #else
-            wxPen pen (wxSystemSettings::GetColour (wxSYS_COLOUR_3DLIGHT ), 1, wxSOLID);
+            wxPen pen (wxSystemSettings::GetColour (wxSYS_COLOUR_3DLIGHT ), 1, wxPENSTYLE_SOLID);
 #endif
             dc.SetPen ((GetBackgroundColour() == *wxWHITE)? pen: *wxWHITE_PEN);
             dc.DrawLine (x_colstart+col_w-1, item->GetY(), x_colstart+col_w-1, item->GetY()+total_h);
@@ -3585,7 +3579,7 @@ void wxTreeListMainWindow::PaintLevel (wxTreeListItem *item, wxDC &dc,
 #if !wxCHECK_VERSION(2, 5, 0)
             wxPen pen (wxSystemSettings::GetSystemColour (wxSYS_COLOUR_3DLIGHT ), 1, wxSOLID);
 #else
-            wxPen pen (wxSystemSettings::GetColour (wxSYS_COLOUR_3DLIGHT ), 1, wxSOLID);
+            wxPen pen (wxSystemSettings::GetColour (wxSYS_COLOUR_3DLIGHT ), 1, wxPENSTYLE_SOLID);
 #endif
             dc.SetPen ((GetBackgroundColour() == *wxWHITE)? pen: *wxWHITE_PEN);
             dc.DrawLine (0, y_top, total_width, y_top);
@@ -3730,7 +3724,7 @@ void wxTreeListMainWindow::OnPaint (wxPaintEvent &WXUNUSED(event)) {
 
     // init device context, clear background (BEFORE changing DC origin...)
     wxAutoBufferedPaintDC dc (this);
-    wxBrush brush(GetBackgroundColour(), wxSOLID);
+    wxBrush brush(GetBackgroundColour(), wxBRUSHSTYLE_SOLID);
     dc.SetBackground(brush);
     dc.Clear();
     DoPrepareDC (dc);
@@ -4468,7 +4462,7 @@ void wxTreeListMainWindow::OnScroll (wxScrollWinEvent& event) {
 
     // TODO
 #if defined(__WXGTK__) && !defined(__WXUNIVERSAL__)
-    wxScrolledWindow::OnScroll(event);
+    wxScrolledWindow::HandleOnScroll(event);
 #else
     HandleOnScroll( event );
 #endif
