diff --git a/CMakeLists.txt b/CMakeLists.txt
new file mode 100644
index 0000000..145f178
--- /dev/null
+++ b/CMakeLists.txt
@@ -0,0 +1,31 @@
+set(lib_name tlc) # wxTreeListCtrl
+#######################################
+# include directories
+include_directories(include)
+#######################################
+# library sources
+########################
+# source
+set(treelist_srcs
+  src/treelistctrl.cpp
+  )
+list(APPEND ${lib_name}_libsrcs ${treelist_srcs})
+########################
+# headers
+set(treelist_hdrs
+  include/wx/treelistctrl.h
+  )
+list(APPEND ${lib_name}_libsrcs ${treelist_hdrs})
+#######################################
+# library
+add_library(${lib_name} STATIC ${${lib_name}_libsrcs})
+set_wxxtarget_properties(${lib_name})
+#######################################
+# install
+install(TARGETS ${lib_name} EXPORT ${PROJECT_NAME}${WX_VER}-targets
+  RUNTIME DESTINATION bin
+  LIBRARY DESTINATION lib
+  ARCHIVE DESTINATION lib
+  )
+install(FILES ${treelist_hdrs} DESTINATION include/${wxinc}/wx/treelistctrl)
+install(EXPORT ${PROJECT_NAME}${WX_VER}-targets DESTINATION lib/cmake)
diff --git a/src/treelistctrl.cpp b/src/treelistctrl.cpp
index 4189e37..29dafd7 100644
--- a/src/treelistctrl.cpp
+++ b/src/treelistctrl.cpp
@@ -50,10 +50,6 @@
 #include <wx/dynarray.h>
 #include <wx/arrimpl.cpp>
 
-#ifdef __WXMAC__
-#include "wx/mac/private.h"
-#endif
-
 #include "wx/treelistctrl.h"
 
 #include <wx/log.h>  // only required for debugging purpose
@@ -2023,8 +2019,6 @@ bool wxTreeListMainWindow::Create (wxTreeListCtrl *parent,
                                    const wxString& name) {
 
 #ifdef __WXMAC__
-    if (style & wxTR_HAS_BUTTONS) style |= wxTR_MAC_BUTTONS;
-    if (style & wxTR_HAS_BUTTONS) style &= ~wxTR_HAS_BUTTONS;
     style &= ~wxTR_LINES_AT_ROOT;
     style |= wxTR_NO_LINES;
 
@@ -4468,7 +4462,7 @@ void wxTreeListMainWindow::OnScroll (wxScrollWinEvent& event) {
 
     // TODO
 #if defined(__WXGTK__) && !defined(__WXUNIVERSAL__)
-    wxScrolledWindow::OnScroll(event);
+    wxScrolledWindow::HandleOnScroll(event);
 #else
     HandleOnScroll( event );
 #endif
